### PR TITLE
feat: Okta authentication support

### PR DESCRIPTION
**Pull request by karenc-bq**
### Summary

New SAML authentication plugin for Okta

#### Added
- new Okta authentication classes
- new SAMLUtils class
- ability to switch to a lighter RDS token implementation that does not require the full RDS SDK in Federated Auth Plugin
- runtime dependency for JSOUP, a Java parsing library. MIT license.

#### Changed
- refactored duplicated helper methods into IamAuthUtils

### By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: examples/AWSDriverExample/src/main/java/software/amazon/OktaAuthPluginExample.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  private static final String CONNECTION_STRING = "jdbc:aws-wrapper:postgresql://database.cluster-xyz.us-east-1.rds.amazonaws.com:5432/postgres";
+
+  public static void main(String[] args) throws SQLException {
+    // Set the Okta Authentication Connection Plugin parameters and the JDBC Wrapper parameters.
+    final Properties properties = new Properties();
+
+    // Enable the Okta Authentication Connection Plugin.
+    properties.setProperty(PropertyDefinition.PLUGINS.name, "okta");
+    properties.setProperty(OktaAuthPlugin.IDP_ENDPOINT.name, "123456789.okta.com");
+    properties.setProperty(OktaAuthPlugin.IDP_NAME.name, "okta");
```
Comment (karenc-bq) [inline examples/AWSDriverExample/src/main/java/software/amazon/OktaAuthPluginExample.java:40]: I don't think we really need the IDP_NAME parameter. All the authentication methods will be enabled and implemented using Plugins so it is a bit weird to have another parameter explicitly stating the IDP name. @sergiyvamz @aaronchung-bitquill thoughts on deprecating it?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/federatedauth/OktaAuthPlugin.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+  public Connection forceConnect(String driverProtocol, HostSpec hostSpec, Properties props,
+      boolean isInitialConnection, JdbcCallable<Connection, SQLException> forceConnectFunc)
+      throws SQLException {
+    return connectInternal(hostSpec, props, forceConnectFunc);
+  }
+
+  private Connection connectInternal(final HostSpec hostSpec, final Properties props,
+      final JdbcCallable<Connection, SQLException> connectFunc) throws SQLException {
+
+    this.samlUtils.checkIdpCredentialsWithFallback(IDP_USERNAME, IDP_PASSWORD, props);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/federatedauth/OktaAuthPlugin.java:155]: I think we need to open telemetry span here. It gives a user a good understanding of authentication timing.

File: examples/AWSDriverExample/src/main/java/software/amazon/OktaAuthPluginExample.java (hunk 1 above)
Comment (sergiyvamz) [reply examples/AWSDriverExample/src/main/java/software/amazon/OktaAuthPluginExample.java:40]: I see that IDP_NAME is used to build a saml url. Is there any chance that IDP_NAME may have value other than "okta"?

**Review by sergiyvamz — APPROVED** [review]

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionPluginManager.java [hunk 3]
```diff
@@ -74,6 +76,8 @@ public class ConnectionPluginManager implements CanReleaseResources, Wrapper {
           put(FailoverConnectionPlugin.class, "plugin:failover");
           put(IamAuthConnectionPlugin.class, "plugin:iam");
           put(AwsSecretsManagerConnectionPlugin.class, "plugin:awsSecretsManager");
+          put(FederatedAuthPlugin.class, "plugin:okta");
```
Comment (aaronchung-bitquill) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionPluginManager.java:79]:
```suggestion
          put(FederatedAuthPlugin.class, "plugin:federatedAuth");
```

File: examples/AWSDriverExample/src/main/java/software/amazon/OktaAuthPluginExample.java (hunk 1 above)
Comment (karenc-bq) [reply examples/AWSDriverExample/src/main/java/software/amazon/OktaAuthPluginExample.java]: Okta does not need idp name in the SAML url so I think we are good

File: wrapper/src/main/java/software/amazon/jdbc/plugin/federatedauth/OktaAuthPluginFactory.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+public class OktaAuthPluginFactory implements ConnectionPluginFactory {
+
+  @Override
+  public ConnectionPlugin getInstance(PluginService pluginService, Properties props) {
+    return new OktaAuthPlugin(pluginService, getCredentialsProviderFactory(pluginService, props));
+  }
+
+  private CredentialsProviderFactory getCredentialsProviderFactory(final PluginService pluginService,
+      final Properties props) {
+    final String idpName = FederatedAuthPlugin.IDP_NAME.getString(props);
```
Comment (aaronchung-bitquill) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/federatedauth/OktaAuthPluginFactory.java:36]:
```suggestion
    final String idpName = OktaAuthPlugin.IDP_NAME.getString(props);
```

File: wrapper/src/main/java/software/amazon/jdbc/plugin/federatedauth/OktaAuthPlugin.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+  public static final AwsWrapperProperty HTTP_CLIENT_SOCKET_TIMEOUT = new AwsWrapperProperty(
+      "httpClientSocketTimeout", String.valueOf(DEFAULT_HTTP_TIMEOUT_MILLIS),
+      "The socket timeout value in milliseconds for the HttpClient used by the OktaAuthPlugin");
+  public static final AwsWrapperProperty HTTP_CLIENT_CONNECT_TIMEOUT = new AwsWrapperProperty(
+      "httpClientConnectTimeout", String.valueOf(DEFAULT_HTTP_TIMEOUT_MILLIS),
+      "The connect timeout value in milliseconds for the HttpClient used by the OktaAuthPlugin");
+  public static final AwsWrapperProperty SSL_INSECURE = new AwsWrapperProperty("sslInsecure", "true",
+      "Whether or not the SSL session is to be secure and the sever's certificates will be verified");
+  public static AwsWrapperProperty
+      IDP_NAME = new AwsWrapperProperty("idpName", "okta", "The name of the Identity Provider implementation used");
```
Comment (aaronchung-bitquill) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/federatedauth/OktaAuthPlugin.java:89]: Based on the prev convo, sound like we can remove `IDP_NAME` and `idpName`.

**Review by aaronchung-bitquill — APPROVED** [review]
Approved!

**Pull request merged**